File: knowledge_inferer/phase2_indicator_calculation.py
import pandas as pd
import numpy as np
import json
from typing import Dict, Any, Optional


class IndicatorCalculator:
    """
    Phase 2: Indicator Calculation
    Tính toán các chỉ số tài chính từ A1 đến D3
    Áp dụng Heuristic H4 - Ưu tiên xác định thuộc tính
    """
    
    def __init__(self, calculation_rules_path: str):
        with open(calculation_rules_path, 'r', encoding='utf-8') as f:
            self.rules = json.load(f)
        
        self.calculation_map = {rule['indicator']: rule for rule in self.rules['calculation_rules']}
    
    def safe_divide(self, a: Any, b: Any) -> float:
        """Chia an toàn, trả về NaN nếu mẫu số = 0 hoặc None"""
        try:
            if pd.isna(a) or pd.isna(b) or b == 0:
                return np.nan
            return float(a) / float(b)
        except:
            return np.nan
    
    def safe_subtract(self, a: Any, b: Any) -> float:
        """Trừ an toàn"""
        try:
            if pd.isna(a) or pd.isna(b):
                return np.nan
            return float(a) - float(b)
        except:
            return np.nan
    
    def safe_add(self, a: Any, b: Any) -> float:
        """Cộng an toàn"""
        try:
            if pd.isna(a) or pd.isna(b):
                return np.nan
            return float(a) + float(b)
        except:
            return np.nan
    
    def safe_abs(self, a: Any) -> float:
        """Giá trị tuyệt đối an toàn"""
        try:
            if pd.isna(a):
                return np.nan
            return abs(float(a))
        except:
            return np.nan
    
    def get_field_value(self, data: pd.Series, field_info: Dict) -> float:
        """Lấy giá trị từ field, áp dụng transform nếu có"""
        field_name = field_info['field']
        value = data.get(field_name, np.nan)
        
        if 'transform' in field_info and field_info['transform'] == 'abs':
            value = self.safe_abs(value)
        
        return value
    
    def calculate_indicator(self, indicator: str, data: pd.Series) -> float:
        """Tính toán một chỉ số cụ thể"""
        if indicator not in self.calculation_map:
            return np.nan
        
        rule = self.calculation_map[indicator]
        calc = rule['calculation']
        
        if calc['operation'] == 'divide':
            numerator_info = calc['numerator']
            denominator_info = calc['denominator']
            
            if 'operation' in numerator_info:
                if numerator_info['operation'] == 'subtract':
                    operands = numerator_info['operands']
                    val1 = self.get_field_value(data, operands[0])
                    val2 = self.get_field_value(data, operands[1])
                    numerator = self.safe_subtract(val1, val2)
                elif numerator_info['operation'] == 'add':
                    operands = numerator_info['operands']
                    val1 = self.get_field_value(data, operands[0])
                    val2 = self.get_field_value(data, operands[1])
                    numerator = self.safe_add(val1, val2)
                else:
                    numerator = np.nan
            else:
                numerator = self.get_field_value(data, numerator_info)
            
            denominator = self.get_field_value(data, denominator_info)
            
            result = self.safe_divide(numerator, denominator)
            
            validation = rule.get('validation', {})
            min_val = validation.get('min_value')
            max_val = validation.get('max_value')
            
            if not pd.isna(result):
                if min_val is not None and result < min_val:
                    pass
                if max_val is not None and result > max_val:
                    pass
            
            return result
        
        return np.nan
    
    def calculate_all_indicators(self, data: pd.Series) -> Dict[str, float]:
        """Tính toán tất cả 12 chỉ số"""
        indicators = {}
        
        for indicator in ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3', 'D1', 'D2', 'D3']:
            indicators[indicator] = self.calculate_indicator(indicator, data)
        
        return indicators
    
    def calculate_batch(self, df: pd.DataFrame) -> pd.DataFrame:
        """Tính toán chỉ số cho toàn bộ DataFrame"""
        print("\n=== Phase 2: Indicator Calculation ===")
        
        result_df = df.copy()
        
        for indicator in ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3', 'D1', 'D2', 'D3']:
            result_df[indicator] = df.apply(
                lambda row: self.calculate_indicator(indicator, row),
                axis=1
            )
            non_null = result_df[indicator].notna().sum()
            print(f"✓ {indicator}: {non_null}/{len(result_df)} calculated")
        
        print("✓ Phase 2 completed\n")
        return result_df
    
    def get_indicator_info(self, indicator: str) -> Optional[Dict]:
        """Lấy thông tin về một chỉ số"""
        return self.calculation_map.get(indicator)

File: src/phases/Phase2IndicatorCalculation.js
/**
 * Phase 2: Indicator Calculation
 * Tính toán các chỉ số tài chính từ A1 đến D3
 * Áp dụng Heuristic H4 - Ưu tiên xác định thuộc tính
 */

import fs from 'fs';

export class IndicatorCalculator {
  constructor(calculationRulesPath) {
    const rulesData = JSON.parse(fs.readFileSync(calculationRulesPath, 'utf-8'));
    this.rules = rulesData;
    this.calculationMap = {};
    
    const rules = rulesData.indicator_rules || rulesData.calculation_rules;
    for (const rule of rules) {
      this.calculationMap[rule.indicator] = rule;
    }
  }

  /**
   * Chia an toàn, trả về null nếu mẫu số = 0 hoặc null
   */
  safeDivide(a, b) {
    if (a === null || a === undefined || b === null || b === undefined || b === 0) {
      return null;
    }
    const numA = parseFloat(a);
    const numB = parseFloat(b);
    if (isNaN(numA) || isNaN(numB) || numB === 0) {
      return null;
    }
    return numA / numB;
  }

  /**
   * Trừ an toàn
   */
  safeSubtract(a, b) {
    if (a === null || a === undefined || b === null || b === undefined) {
      return null;
    }
    const numA = parseFloat(a);
    const numB = parseFloat(b);
    if (isNaN(numA) || isNaN(numB)) {
      return null;
    }
    return numA - numB;
  }

  /**
   * Cộng an toàn
   */
  safeAdd(a, b) {
    if (a === null || a === undefined || b === null || b === undefined) {
      return null;
    }
    const numA = parseFloat(a);
    const numB = parseFloat(b);
    if (isNaN(numA) || isNaN(numB)) {
      return null;
    }
    return numA + numB;
  }

  /**
   * Giá trị tuyệt đối an toàn
   */
  safeAbs(a) {
    if (a === null || a === undefined) {
      return null;
    }
    const num = parseFloat(a);
    if (isNaN(num)) {
      return null;
    }
    return Math.abs(num);
  }

  /**
   * Lấy giá trị từ field, áp dụng transform nếu có
   */
  getFieldValue(data, fieldInfo) {
    const fieldName = fieldInfo.field;
    let value = data[fieldName];

    if (value === undefined || value === null || value === '') {
      return null;
    }

    value = parseFloat(value);
    if (isNaN(value)) {
      return null;
    }

    if (fieldInfo.transform === 'abs') {
      value = Math.abs(value);
    }

    return value;
  }

  /**
   * Tính toán một chỉ số cụ thể
   */
  calculateIndicator(indicator, data) {
    if (!this.calculationMap[indicator]) {
      return null;
    }

    const rule = this.calculationMap[indicator];
    const calc = rule.calculation;

    if (calc.operation === 'divide') {
      const numeratorInfo = calc.numerator;
      const denominatorInfo = calc.denominator;

      let numerator;
      let denominator;

      // Calculate numerator
      if (numeratorInfo.operation === 'subtract') {
        const operands = numeratorInfo.operands;
        const val1 = this.getFieldValue(data, operands[0]);
        const val2 = this.getFieldValue(data, operands[1]);
        numerator = this.safeSubtract(val1, val2);
      } else if (numeratorInfo.operation === 'add') {
        const operands = numeratorInfo.operands;
        const val1 = this.getFieldValue(data, operands[0]);
        const val2 = this.getFieldValue(data, operands[1]);
        numerator = this.safeAdd(val1, val2);
      } else {
        numerator = this.getFieldValue(data, numeratorInfo);
      }

      // Calculate denominator
      if (denominatorInfo.operation === 'subtract') {
        const operands = denominatorInfo.operands;
        const val1 = this.getFieldValue(data, operands[0]);
        const val2 = this.getFieldValue(data, operands[1]);
        denominator = this.safeSubtract(val1, val2);
      } else if (denominatorInfo.operation === 'add') {
        const operands = denominatorInfo.operands;
        const val1 = this.getFieldValue(data, operands[0]);
        const val2 = this.getFieldValue(data, operands[1]);
        denominator = this.safeAdd(val1, val2);
      } else {
        denominator = this.getFieldValue(data, denominatorInfo);
      }

      return this.safeDivide(numerator, denominator);
    }

    return null;
  }

  /**
   * Tính toán tất cả chỉ số cho một công ty
   */
  calculateAll(data) {

    const indicators = {};

    for (const indicator of Object.keys(this.calculationMap)) {
      const value = this.calculateIndicator(indicator, data);
      indicators[indicator] = value;

      if (value !== null) {
      } else {
      }
    }

    return indicators;
  }

  /**
   * Lấy thông tin về một chỉ số
   */
  getIndicatorInfo(indicator) {
    return this.calculationMap[indicator] || null;
  }
}
